{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json\n",
    "import math"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Directories of the TCGA data\n",
    "CLIN_PATH = '../data/clinical.project-TCGA-BRCA.2016-09-30T12_32_20.666404.json' #Patient data\n",
    "DIST_PATH = '../data/dd5_5_for_TCGA.txt' # Distance metric\n",
    "GENOME_PATH = '../data/BRCA.rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.data.txt' #Genome data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Reading data files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Clinical data of patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Clinical data of patients\n",
    "big_df = pd.read_json(CLIN_PATH)\n",
    "\n",
    "# Change lists to dictionaries to easily expand to columns in DataFrame\n",
    "def dictify(df):\n",
    "    df['diagnoses'] = df['diagnoses'][0]\n",
    "    df['exposures'] = df['exposures'][0]\n",
    "    return df\n",
    "\n",
    "big_df = big_df.dropna().apply(dictify,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Expand dictionaries to columns in DataFrame\n",
    "df_clin = pd.DataFrame(big_df['case_id']).join(big_df['demographic'].apply(pd.Series),how='outer',lsuffix='_case_id',rsuffix='_demo')\\\n",
    "            .join(big_df['diagnoses'].apply(pd.Series),how='outer',lsuffix = '_demo',rsuffix='_diag')\\\n",
    "            .join(big_df['exposures'].apply(pd.Series),how='outer',lsuffix = '_diag',rsuffix='_expo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Drop nulls rows and columns\n",
    "df_clin.dropna(how='all', inplace = True)\n",
    "df_clin.dropna(how='all', inplace = True, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>demographic_id</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>submitter_id_demo</th>\n",
       "      <th>updated_datetime_demo</th>\n",
       "      <th>year_of_birth</th>\n",
       "      <th>year_of_death</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>...</th>\n",
       "      <th>site_of_resection_or_biopsy</th>\n",
       "      <th>submitter_id_diag</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>updated_datetime_diag</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>exposure_id</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>updated_datetime</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b205c89f-af62-4186-acad-ed23d243fa98</td>\n",
       "      <td>55ae5a3b-e509-5fcd-98c9-c5007b448890</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-A2-A0YL_demographic</td>\n",
       "      <td>2016-09-02T19:00:43.039678-05:00</td>\n",
       "      <td>1962.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17702.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-A2-A0YL_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iiia</td>\n",
       "      <td>2016-09-02T19:00:43.039678-05:00</td>\n",
       "      <td>alive</td>\n",
       "      <td>23b6a296-e149-5801-aa20-e144ef683abe</td>\n",
       "      <td>TCGA-A2-A0YL_exposure</td>\n",
       "      <td>2016-09-02T19:00:43.039678-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>70f34c5c-3671-44c8-9469-99f9786efec1</td>\n",
       "      <td>498e27fb-2201-524a-bdd1-86f4e53dc54b</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-D8-A1XR_demographic</td>\n",
       "      <td>2016-09-02T19:07:03.114741-05:00</td>\n",
       "      <td>1954.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20488.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-D8-A1XR_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iib</td>\n",
       "      <td>2016-09-02T19:07:03.114741-05:00</td>\n",
       "      <td>alive</td>\n",
       "      <td>6dfa11dc-4266-57c9-b746-02a6c7bb9d23</td>\n",
       "      <td>TCGA-D8-A1XR_exposure</td>\n",
       "      <td>2016-09-02T19:07:03.114741-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8986a141-eae7-4157-b695-02cc6fc3b071</td>\n",
       "      <td>8faa57d4-1199-53ce-a0ce-24d358841153</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-BH-A1ET_demographic</td>\n",
       "      <td>2016-09-02T18:57:52.265307-05:00</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>20425.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-BH-A1ET_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage i</td>\n",
       "      <td>2016-09-02T18:57:52.265307-05:00</td>\n",
       "      <td>dead</td>\n",
       "      <td>e72e6b02-58be-539a-af7f-f2714f14899a</td>\n",
       "      <td>TCGA-BH-A1ET_exposure</td>\n",
       "      <td>2016-09-02T18:57:52.265307-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7317e605-93ba-49a9-8743-20613b02767a</td>\n",
       "      <td>4394b44c-b51d-5b96-b62b-db7f49007b43</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-E2-A1LE_demographic</td>\n",
       "      <td>2016-09-02T19:09:17.163541-05:00</td>\n",
       "      <td>1937.0</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>26239.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-E2-A1LE_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iiic</td>\n",
       "      <td>2016-09-02T19:09:17.163541-05:00</td>\n",
       "      <td>dead</td>\n",
       "      <td>4c456a08-ffda-507f-a3e3-d273e5c49864</td>\n",
       "      <td>TCGA-E2-A1LE_exposure</td>\n",
       "      <td>2016-09-02T19:09:17.163541-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7e1673f8-5758-4963-8804-d5e39f06205b</td>\n",
       "      <td>eebe9198-74c8-5996-bf88-c56d5fd6832f</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>TCGA-C8-A12V_demographic</td>\n",
       "      <td>2016-09-02T19:02:30.147198-05:00</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20346.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-C8-A12V_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iia</td>\n",
       "      <td>2016-09-02T19:02:30.147198-05:00</td>\n",
       "      <td>alive</td>\n",
       "      <td>dd5fbf53-64a4-5bc0-8ca7-ae05abbcbbac</td>\n",
       "      <td>TCGA-C8-A12V_exposure</td>\n",
       "      <td>2016-09-02T19:02:30.147198-05:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id                        demographic_id  \\\n",
       "0  b205c89f-af62-4186-acad-ed23d243fa98  55ae5a3b-e509-5fcd-98c9-c5007b448890   \n",
       "1  70f34c5c-3671-44c8-9469-99f9786efec1  498e27fb-2201-524a-bdd1-86f4e53dc54b   \n",
       "2  8986a141-eae7-4157-b695-02cc6fc3b071  8faa57d4-1199-53ce-a0ce-24d358841153   \n",
       "3  7317e605-93ba-49a9-8743-20613b02767a  4394b44c-b51d-5b96-b62b-db7f49007b43   \n",
       "4  7e1673f8-5758-4963-8804-d5e39f06205b  eebe9198-74c8-5996-bf88-c56d5fd6832f   \n",
       "\n",
       "                ethnicity  gender   race         submitter_id_demo  \\\n",
       "0  not hispanic or latino  female  white  TCGA-A2-A0YL_demographic   \n",
       "1  not hispanic or latino  female  white  TCGA-D8-A1XR_demographic   \n",
       "2  not hispanic or latino  female  white  TCGA-BH-A1ET_demographic   \n",
       "3  not hispanic or latino  female  white  TCGA-E2-A1LE_demographic   \n",
       "4  not hispanic or latino  female  asian  TCGA-C8-A12V_demographic   \n",
       "\n",
       "              updated_datetime_demo  year_of_birth  year_of_death  \\\n",
       "0  2016-09-02T19:00:43.039678-05:00         1962.0            NaN   \n",
       "1  2016-09-02T19:07:03.114741-05:00         1954.0            NaN   \n",
       "2  2016-09-02T18:57:52.265307-05:00         1944.0         2005.0   \n",
       "3  2016-09-02T19:09:17.163541-05:00         1937.0         2010.0   \n",
       "4  2016-09-02T19:02:30.147198-05:00         1955.0            NaN   \n",
       "\n",
       "   age_at_diagnosis                ...                 \\\n",
       "0           17702.0                ...                  \n",
       "1           20488.0                ...                  \n",
       "2           20425.0                ...                  \n",
       "3           26239.0                ...                  \n",
       "4           20346.0                ...                  \n",
       "\n",
       "  site_of_resection_or_biopsy       submitter_id_diag  \\\n",
       "0                       c50.9  TCGA-A2-A0YL_diagnosis   \n",
       "1                       c50.9  TCGA-D8-A1XR_diagnosis   \n",
       "2                       c50.9  TCGA-BH-A1ET_diagnosis   \n",
       "3                       c50.9  TCGA-E2-A1LE_diagnosis   \n",
       "4                       c50.9  TCGA-C8-A12V_diagnosis   \n",
       "\n",
       "   tissue_or_organ_of_origin   tumor_grade tumor_stage  \\\n",
       "0                      c50.9  not reported  stage iiia   \n",
       "1                      c50.9  not reported   stage iib   \n",
       "2                      c50.9  not reported     stage i   \n",
       "3                      c50.9  not reported  stage iiic   \n",
       "4                      c50.9  not reported   stage iia   \n",
       "\n",
       "              updated_datetime_diag vital_status  \\\n",
       "0  2016-09-02T19:00:43.039678-05:00        alive   \n",
       "1  2016-09-02T19:07:03.114741-05:00        alive   \n",
       "2  2016-09-02T18:57:52.265307-05:00         dead   \n",
       "3  2016-09-02T19:09:17.163541-05:00         dead   \n",
       "4  2016-09-02T19:02:30.147198-05:00        alive   \n",
       "\n",
       "                            exposure_id           submitter_id  \\\n",
       "0  23b6a296-e149-5801-aa20-e144ef683abe  TCGA-A2-A0YL_exposure   \n",
       "1  6dfa11dc-4266-57c9-b746-02a6c7bb9d23  TCGA-D8-A1XR_exposure   \n",
       "2  e72e6b02-58be-539a-af7f-f2714f14899a  TCGA-BH-A1ET_exposure   \n",
       "3  4c456a08-ffda-507f-a3e3-d273e5c49864  TCGA-E2-A1LE_exposure   \n",
       "4  dd5fbf53-64a4-5bc0-8ca7-ae05abbcbbac  TCGA-C8-A12V_exposure   \n",
       "\n",
       "                   updated_datetime  \n",
       "0  2016-09-02T19:00:43.039678-05:00  \n",
       "1  2016-09-02T19:07:03.114741-05:00  \n",
       "2  2016-09-02T18:57:52.265307-05:00  \n",
       "3  2016-09-02T19:09:17.163541-05:00  \n",
       "4  2016-09-02T19:02:30.147198-05:00  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clin.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distances matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Creating DataFrame from the distance-correlation matrix data\n",
    "df_distances = pd.read_csv(DIST_PATH,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA.3C.AAAU.0.Dis</th>\n",
       "      <th>TCGA.3C.AALI.0.Dis</th>\n",
       "      <th>TCGA.3C.AALJ.0.Dis</th>\n",
       "      <th>TCGA.3C.AALK.0.Dis</th>\n",
       "      <th>TCGA.4H.AAAK.0.Dis</th>\n",
       "      <th>TCGA.5L.AAT0.0.Dis</th>\n",
       "      <th>TCGA.5L.AAT1.0.Dis</th>\n",
       "      <th>TCGA.5T.A9QA.0.Dis</th>\n",
       "      <th>TCGA.A1.A0SB.0.Dis</th>\n",
       "      <th>TCGA.A1.A0SD.0.Dis</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA.UL.AAZ6.0.Dis</th>\n",
       "      <th>TCGA.UU.A93S.0.Dis</th>\n",
       "      <th>TCGA.V7.A7HQ.0.Dis</th>\n",
       "      <th>TCGA.W8.A86G.0.Dis</th>\n",
       "      <th>TCGA.WT.AB41.0.Dis</th>\n",
       "      <th>TCGA.WT.AB44.0.Dis</th>\n",
       "      <th>TCGA.XX.A899.0.Dis</th>\n",
       "      <th>TCGA.XX.A89A.0.Dis</th>\n",
       "      <th>TCGA.Z7.A8R5.0.Dis</th>\n",
       "      <th>TCGA.Z7.A8R6.0.Dis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AAAU.0.Dis</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.120880</td>\n",
       "      <td>0.105801</td>\n",
       "      <td>0.109540</td>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.116020</td>\n",
       "      <td>0.138326</td>\n",
       "      <td>0.139012</td>\n",
       "      <td>0.164870</td>\n",
       "      <td>0.109789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130849</td>\n",
       "      <td>0.177643</td>\n",
       "      <td>0.202318</td>\n",
       "      <td>0.118387</td>\n",
       "      <td>0.204810</td>\n",
       "      <td>0.172970</td>\n",
       "      <td>0.131472</td>\n",
       "      <td>0.147361</td>\n",
       "      <td>0.160945</td>\n",
       "      <td>0.102312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AALI.0.Dis</th>\n",
       "      <td>0.120880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.105240</td>\n",
       "      <td>0.102997</td>\n",
       "      <td>0.095894</td>\n",
       "      <td>0.142688</td>\n",
       "      <td>0.183251</td>\n",
       "      <td>0.114400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117577</td>\n",
       "      <td>0.143124</td>\n",
       "      <td>0.172721</td>\n",
       "      <td>0.125615</td>\n",
       "      <td>0.158141</td>\n",
       "      <td>0.134712</td>\n",
       "      <td>0.116518</td>\n",
       "      <td>0.119447</td>\n",
       "      <td>0.129105</td>\n",
       "      <td>0.101502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AALJ.0.Dis</th>\n",
       "      <td>0.105801</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.072653</td>\n",
       "      <td>0.091719</td>\n",
       "      <td>0.082123</td>\n",
       "      <td>0.096455</td>\n",
       "      <td>0.141006</td>\n",
       "      <td>0.181756</td>\n",
       "      <td>0.102686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108979</td>\n",
       "      <td>0.147735</td>\n",
       "      <td>0.151162</td>\n",
       "      <td>0.113216</td>\n",
       "      <td>0.138139</td>\n",
       "      <td>0.120631</td>\n",
       "      <td>0.100941</td>\n",
       "      <td>0.103558</td>\n",
       "      <td>0.112343</td>\n",
       "      <td>0.083557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AALK.0.Dis</th>\n",
       "      <td>0.109540</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.072653</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040750</td>\n",
       "      <td>0.040376</td>\n",
       "      <td>0.062932</td>\n",
       "      <td>0.144121</td>\n",
       "      <td>0.129292</td>\n",
       "      <td>0.053960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117827</td>\n",
       "      <td>0.158452</td>\n",
       "      <td>0.146738</td>\n",
       "      <td>0.062870</td>\n",
       "      <td>0.165742</td>\n",
       "      <td>0.102312</td>\n",
       "      <td>0.060814</td>\n",
       "      <td>0.073400</td>\n",
       "      <td>0.080130</td>\n",
       "      <td>0.070721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.4H.AAAK.0.Dis</th>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.105240</td>\n",
       "      <td>0.091719</td>\n",
       "      <td>0.040750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037074</td>\n",
       "      <td>0.065612</td>\n",
       "      <td>0.152471</td>\n",
       "      <td>0.113901</td>\n",
       "      <td>0.054209</td>\n",
       "      <td>...</td>\n",
       "      <td>0.143560</td>\n",
       "      <td>0.180884</td>\n",
       "      <td>0.172347</td>\n",
       "      <td>0.065300</td>\n",
       "      <td>0.187800</td>\n",
       "      <td>0.112343</td>\n",
       "      <td>0.059381</td>\n",
       "      <td>0.069163</td>\n",
       "      <td>0.097015</td>\n",
       "      <td>0.078198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1093 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    TCGA.3C.AAAU.0.Dis  TCGA.3C.AALI.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.000000            0.120880   \n",
       "TCGA.3C.AALI.0.Dis            0.120880            0.000000   \n",
       "TCGA.3C.AALJ.0.Dis            0.105801            0.086734   \n",
       "TCGA.3C.AALK.0.Dis            0.109540            0.086734   \n",
       "TCGA.4H.AAAK.0.Dis            0.118014            0.105240   \n",
       "\n",
       "                    TCGA.3C.AALJ.0.Dis  TCGA.3C.AALK.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.105801            0.109540   \n",
       "TCGA.3C.AALI.0.Dis            0.086734            0.086734   \n",
       "TCGA.3C.AALJ.0.Dis            0.000000            0.072653   \n",
       "TCGA.3C.AALK.0.Dis            0.072653            0.000000   \n",
       "TCGA.4H.AAAK.0.Dis            0.091719            0.040750   \n",
       "\n",
       "                    TCGA.4H.AAAK.0.Dis  TCGA.5L.AAT0.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.118014            0.116020   \n",
       "TCGA.3C.AALI.0.Dis            0.105240            0.102997   \n",
       "TCGA.3C.AALJ.0.Dis            0.091719            0.082123   \n",
       "TCGA.3C.AALK.0.Dis            0.040750            0.040376   \n",
       "TCGA.4H.AAAK.0.Dis            0.000000            0.037074   \n",
       "\n",
       "                    TCGA.5L.AAT1.0.Dis  TCGA.5T.A9QA.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.138326            0.139012   \n",
       "TCGA.3C.AALI.0.Dis            0.095894            0.142688   \n",
       "TCGA.3C.AALJ.0.Dis            0.096455            0.141006   \n",
       "TCGA.3C.AALK.0.Dis            0.062932            0.144121   \n",
       "TCGA.4H.AAAK.0.Dis            0.065612            0.152471   \n",
       "\n",
       "                    TCGA.A1.A0SB.0.Dis  TCGA.A1.A0SD.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.164870            0.109789   \n",
       "TCGA.3C.AALI.0.Dis            0.183251            0.114400   \n",
       "TCGA.3C.AALJ.0.Dis            0.181756            0.102686   \n",
       "TCGA.3C.AALK.0.Dis            0.129292            0.053960   \n",
       "TCGA.4H.AAAK.0.Dis            0.113901            0.054209   \n",
       "\n",
       "                           ...          TCGA.UL.AAZ6.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis         ...                    0.130849   \n",
       "TCGA.3C.AALI.0.Dis         ...                    0.117577   \n",
       "TCGA.3C.AALJ.0.Dis         ...                    0.108979   \n",
       "TCGA.3C.AALK.0.Dis         ...                    0.117827   \n",
       "TCGA.4H.AAAK.0.Dis         ...                    0.143560   \n",
       "\n",
       "                    TCGA.UU.A93S.0.Dis  TCGA.V7.A7HQ.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.177643            0.202318   \n",
       "TCGA.3C.AALI.0.Dis            0.143124            0.172721   \n",
       "TCGA.3C.AALJ.0.Dis            0.147735            0.151162   \n",
       "TCGA.3C.AALK.0.Dis            0.158452            0.146738   \n",
       "TCGA.4H.AAAK.0.Dis            0.180884            0.172347   \n",
       "\n",
       "                    TCGA.W8.A86G.0.Dis  TCGA.WT.AB41.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.118387            0.204810   \n",
       "TCGA.3C.AALI.0.Dis            0.125615            0.158141   \n",
       "TCGA.3C.AALJ.0.Dis            0.113216            0.138139   \n",
       "TCGA.3C.AALK.0.Dis            0.062870            0.165742   \n",
       "TCGA.4H.AAAK.0.Dis            0.065300            0.187800   \n",
       "\n",
       "                    TCGA.WT.AB44.0.Dis  TCGA.XX.A899.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.172970            0.131472   \n",
       "TCGA.3C.AALI.0.Dis            0.134712            0.116518   \n",
       "TCGA.3C.AALJ.0.Dis            0.120631            0.100941   \n",
       "TCGA.3C.AALK.0.Dis            0.102312            0.060814   \n",
       "TCGA.4H.AAAK.0.Dis            0.112343            0.059381   \n",
       "\n",
       "                    TCGA.XX.A89A.0.Dis  TCGA.Z7.A8R5.0.Dis  TCGA.Z7.A8R6.0.Dis  \n",
       "TCGA.3C.AAAU.0.Dis            0.147361            0.160945            0.102312  \n",
       "TCGA.3C.AALI.0.Dis            0.119447            0.129105            0.101502  \n",
       "TCGA.3C.AALJ.0.Dis            0.103558            0.112343            0.083557  \n",
       "TCGA.3C.AALK.0.Dis            0.073400            0.080130            0.070721  \n",
       "TCGA.4H.AAAK.0.Dis            0.069163            0.097015            0.078198  \n",
       "\n",
       "[5 rows x 1093 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_distances.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Genome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Creating DataFrame from the Genome data\n",
    "#Note please change the first two words of the text file: \"Hybridization REF\" to \"Hybridization_REF\"\n",
    "with open(GENOME_PATH) as f:\n",
    "    row = [row.split()[0:] for row in f]\n",
    "df_genome = pd.DataFrame(row[1:],columns=row[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hybridization_REF</th>\n",
       "      <th>TCGA-3C-AAAU-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALI-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALJ-01A-31R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALK-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-4H-AAAK-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5L-AAT0-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5L-AAT1-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5T-A9QA-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-A1-A0SB-01A-11R-A144-07</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-UL-AAZ6-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-UU-A93S-01A-21R-A41B-07</th>\n",
       "      <th>TCGA-V7-A7HQ-01A-11R-A33J-07</th>\n",
       "      <th>TCGA-W8-A86G-01A-21R-A36F-07</th>\n",
       "      <th>TCGA-WT-AB41-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-WT-AB44-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-XX-A899-01A-11R-A36F-07</th>\n",
       "      <th>TCGA-XX-A89A-01A-11R-A36F-07</th>\n",
       "      <th>TCGA-Z7-A8R5-01A-42R-A41B-07</th>\n",
       "      <th>TCGA-Z7-A8R6-01A-11R-A41B-07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>...</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>?|100130426</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9066</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>?|100133144</td>\n",
       "      <td>16.3644</td>\n",
       "      <td>9.2659</td>\n",
       "      <td>11.6228</td>\n",
       "      <td>12.0894</td>\n",
       "      <td>6.8468</td>\n",
       "      <td>3.9889</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4644</td>\n",
       "      <td>15.3396</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3992</td>\n",
       "      <td>4.3126</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>5.5624</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>14.3858</td>\n",
       "      <td>22.3240</td>\n",
       "      <td>2.2638</td>\n",
       "      <td>6.8865</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1213 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hybridization_REF TCGA-3C-AAAU-01A-11R-A41B-07 TCGA-3C-AALI-01A-11R-A41B-07  \\\n",
       "0           gene_id             normalized_count             normalized_count   \n",
       "1       ?|100130426                       0.0000                       0.0000   \n",
       "2       ?|100133144                      16.3644                       9.2659   \n",
       "\n",
       "  TCGA-3C-AALJ-01A-31R-A41B-07 TCGA-3C-AALK-01A-11R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.9066                       0.0000   \n",
       "2                      11.6228                      12.0894   \n",
       "\n",
       "  TCGA-4H-AAAK-01A-12R-A41B-07 TCGA-5L-AAT0-01A-12R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       6.8468                       3.9889   \n",
       "\n",
       "  TCGA-5L-AAT1-01A-12R-A41B-07 TCGA-5T-A9QA-01A-11R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.0000                       1.4644   \n",
       "\n",
       "  TCGA-A1-A0SB-01A-11R-A144-07             ...               \\\n",
       "0             normalized_count             ...                \n",
       "1                       0.0000             ...                \n",
       "2                      15.3396             ...                \n",
       "\n",
       "  TCGA-UL-AAZ6-01A-11R-A41B-07 TCGA-UU-A93S-01A-21R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.3992                       4.3126   \n",
       "\n",
       "  TCGA-V7-A7HQ-01A-11R-A33J-07 TCGA-W8-A86G-01A-21R-A36F-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.0000                       5.5624   \n",
       "\n",
       "  TCGA-WT-AB41-01A-11R-A41B-07 TCGA-WT-AB44-01A-11R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.0000                       0.0000   \n",
       "\n",
       "  TCGA-XX-A899-01A-11R-A36F-07 TCGA-XX-A89A-01A-11R-A36F-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                      14.3858                      22.3240   \n",
       "\n",
       "  TCGA-Z7-A8R5-01A-42R-A41B-07 TCGA-Z7-A8R6-01A-11R-A41B-07  \n",
       "0             normalized_count             normalized_count  \n",
       "1                       0.0000                       0.0000  \n",
       "2                       2.2638                       6.8865  \n",
       "\n",
       "[3 rows x 1213 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genome.head(n=3)\n",
    "#Row 0 as it has no information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cleaning Data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Removing data of the patients that are not in the three datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of references: 1093\n"
     ]
    }
   ],
   "source": [
    "#We know that the label \"TCGA-XX-XXXX\" is a reference for each patient\n",
    "#We take as a reference the file with less patients, that is to say: distances matrix data\n",
    "\n",
    "#Taking df_distances as reference to match the other datasets \n",
    "ref_distance = [val.replace(\".\",\"-\")[0:12] for val in df_distances.columns.tolist()]\n",
    "print(\"Number of references: \" + str(len(ref_distance)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of references: 1097\n",
      "{'TCGA-AC-A5EI', 'TCGA-C8-A9FZ', 'TCGA-AR-A0U1', 'TCGA-A7-A0DC'}\n"
     ]
    }
   ],
   "source": [
    "#Getting the references for clinical data\n",
    "ref_clinical = [val[0:12] for val in df_clin['submitter_id_demo'].astype(str).tolist()]\n",
    "print(\"Number of references: \" + str(len(ref_clinical)))\n",
    "\n",
    "#Getting the references that are not in the matrix data (References that need to be removed)\n",
    "ref_clinical_to_remove=list(set(ref_clinical)-set(ref_distance))\n",
    "print(set(ref_clinical)-set(ref_distance))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of references: 1212\n"
     ]
    }
   ],
   "source": [
    "#Getting the references for genome data\n",
    "ref_genome = [val[0:12]for val in df_genome.columns.tolist()[1:]]\n",
    "print(\"Number of references: \" + str(len(ref_genome)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1093"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#References containing only primary solid tumor TP (01)\n",
    "ref_genome_tumor=[val for val in df_genome.columns.tolist()[1:] if val[13:15]==\"01\"]\n",
    "len(ref_genome_tumor)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7\n"
     ]
    }
   ],
   "source": [
    "#References containing only metastatic TM (06)\n",
    "ref_genome_tumor_06=[val for val in df_genome.columns.tolist()[1:] if val[13:15]==\"06\"]\n",
    "print(len(ref_genome_tumor_06))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "112\n"
     ]
    }
   ],
   "source": [
    "#References containing solid tissue normal NT (11) -\n",
    "ref_genome_no_tumor=[val for val in df_genome.columns.tolist()[1:] if val[13:15]==\"11\"]\n",
    "print(len(ref_genome_no_tumor))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Creating new genome DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Since we are only going to focus on the primary solid tumor (01), we can get rid of the other references with 11 and 06\n",
    "df_genome_new=df_genome[[\"Hybridization_REF\"]+[val for val in df_genome.columns.tolist()[1:] if val[13:15]==\"01\"]].iloc[1:].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "df_genome_new.reset_index(drop=True,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Splitting the Hybridization_REF in order to have gene_id and gene_id_reference\n",
    "list_gene = [val.split(\"|\")[0] for val in df_genome_new[\"Hybridization_REF\"].tolist()]\n",
    "list_gene_id = [val.split(\"|\")[1] for val in df_genome_new[\"Hybridization_REF\"].tolist()]\n",
    "df_genome_new[\"gene\"] = list_gene\n",
    "df_genome_new[\"gene_id\"] = list_gene_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "cols = df_genome_new.columns.tolist()\n",
    "df_genome_new = df_genome_new[[cols[-2]]+[cols[-1]]+cols[:-2]]\n",
    "del df_genome_new[\"Hybridization_REF\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>TCGA-3C-AAAU-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALI-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALJ-01A-31R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALK-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-4H-AAAK-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5L-AAT0-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5L-AAT1-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5T-A9QA-01A-11R-A41B-07</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-UL-AAZ6-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-UU-A93S-01A-21R-A41B-07</th>\n",
       "      <th>TCGA-V7-A7HQ-01A-11R-A33J-07</th>\n",
       "      <th>TCGA-W8-A86G-01A-21R-A36F-07</th>\n",
       "      <th>TCGA-WT-AB41-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-WT-AB44-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-XX-A899-01A-11R-A36F-07</th>\n",
       "      <th>TCGA-XX-A89A-01A-11R-A36F-07</th>\n",
       "      <th>TCGA-Z7-A8R5-01A-42R-A41B-07</th>\n",
       "      <th>TCGA-Z7-A8R6-01A-11R-A41B-07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>?</td>\n",
       "      <td>100130426</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9066</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>?</td>\n",
       "      <td>100133144</td>\n",
       "      <td>16.3644</td>\n",
       "      <td>9.2659</td>\n",
       "      <td>11.6228</td>\n",
       "      <td>12.0894</td>\n",
       "      <td>6.8468</td>\n",
       "      <td>3.9889</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4644</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3992</td>\n",
       "      <td>4.3126</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>5.5624</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>14.3858</td>\n",
       "      <td>22.3240</td>\n",
       "      <td>2.2638</td>\n",
       "      <td>6.8865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>?</td>\n",
       "      <td>100134869</td>\n",
       "      <td>12.9316</td>\n",
       "      <td>17.3790</td>\n",
       "      <td>9.2294</td>\n",
       "      <td>11.0799</td>\n",
       "      <td>14.4298</td>\n",
       "      <td>13.6090</td>\n",
       "      <td>10.5949</td>\n",
       "      <td>8.9958</td>\n",
       "      <td>...</td>\n",
       "      <td>14.3720</td>\n",
       "      <td>10.8828</td>\n",
       "      <td>3.0792</td>\n",
       "      <td>14.3711</td>\n",
       "      <td>6.3091</td>\n",
       "      <td>3.2580</td>\n",
       "      <td>21.4409</td>\n",
       "      <td>27.2744</td>\n",
       "      <td>7.2933</td>\n",
       "      <td>24.7795</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>?</td>\n",
       "      <td>10357</td>\n",
       "      <td>52.1503</td>\n",
       "      <td>69.7553</td>\n",
       "      <td>154.2974</td>\n",
       "      <td>143.8643</td>\n",
       "      <td>84.2128</td>\n",
       "      <td>114.2572</td>\n",
       "      <td>115.9984</td>\n",
       "      <td>107.5628</td>\n",
       "      <td>...</td>\n",
       "      <td>135.6241</td>\n",
       "      <td>136.1288</td>\n",
       "      <td>29.9974</td>\n",
       "      <td>128.3151</td>\n",
       "      <td>53.6278</td>\n",
       "      <td>42.2643</td>\n",
       "      <td>137.7756</td>\n",
       "      <td>64.1427</td>\n",
       "      <td>85.0461</td>\n",
       "      <td>167.5511</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>?</td>\n",
       "      <td>10431</td>\n",
       "      <td>408.0760</td>\n",
       "      <td>563.8934</td>\n",
       "      <td>1360.8341</td>\n",
       "      <td>865.5358</td>\n",
       "      <td>766.3830</td>\n",
       "      <td>807.7431</td>\n",
       "      <td>1108.3945</td>\n",
       "      <td>1420.5021</td>\n",
       "      <td>...</td>\n",
       "      <td>1570.1445</td>\n",
       "      <td>2886.3965</td>\n",
       "      <td>1721.8816</td>\n",
       "      <td>697.6744</td>\n",
       "      <td>1245.2681</td>\n",
       "      <td>1877.4180</td>\n",
       "      <td>652.7559</td>\n",
       "      <td>722.7208</td>\n",
       "      <td>1140.2801</td>\n",
       "      <td>1003.5668</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1095 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene    gene_id TCGA-3C-AAAU-01A-11R-A41B-07 TCGA-3C-AALI-01A-11R-A41B-07  \\\n",
       "0    ?  100130426                       0.0000                       0.0000   \n",
       "1    ?  100133144                      16.3644                       9.2659   \n",
       "2    ?  100134869                      12.9316                      17.3790   \n",
       "3    ?      10357                      52.1503                      69.7553   \n",
       "4    ?      10431                     408.0760                     563.8934   \n",
       "\n",
       "  TCGA-3C-AALJ-01A-31R-A41B-07 TCGA-3C-AALK-01A-11R-A41B-07  \\\n",
       "0                       0.9066                       0.0000   \n",
       "1                      11.6228                      12.0894   \n",
       "2                       9.2294                      11.0799   \n",
       "3                     154.2974                     143.8643   \n",
       "4                    1360.8341                     865.5358   \n",
       "\n",
       "  TCGA-4H-AAAK-01A-12R-A41B-07 TCGA-5L-AAT0-01A-12R-A41B-07  \\\n",
       "0                       0.0000                       0.0000   \n",
       "1                       6.8468                       3.9889   \n",
       "2                      14.4298                      13.6090   \n",
       "3                      84.2128                     114.2572   \n",
       "4                     766.3830                     807.7431   \n",
       "\n",
       "  TCGA-5L-AAT1-01A-12R-A41B-07 TCGA-5T-A9QA-01A-11R-A41B-07  \\\n",
       "0                       0.0000                       0.0000   \n",
       "1                       0.0000                       1.4644   \n",
       "2                      10.5949                       8.9958   \n",
       "3                     115.9984                     107.5628   \n",
       "4                    1108.3945                    1420.5021   \n",
       "\n",
       "              ...              TCGA-UL-AAZ6-01A-11R-A41B-07  \\\n",
       "0             ...                                    0.0000   \n",
       "1             ...                                    0.3992   \n",
       "2             ...                                   14.3720   \n",
       "3             ...                                  135.6241   \n",
       "4             ...                                 1570.1445   \n",
       "\n",
       "  TCGA-UU-A93S-01A-21R-A41B-07 TCGA-V7-A7HQ-01A-11R-A33J-07  \\\n",
       "0                       0.0000                       0.0000   \n",
       "1                       4.3126                       0.0000   \n",
       "2                      10.8828                       3.0792   \n",
       "3                     136.1288                      29.9974   \n",
       "4                    2886.3965                    1721.8816   \n",
       "\n",
       "  TCGA-W8-A86G-01A-21R-A36F-07 TCGA-WT-AB41-01A-11R-A41B-07  \\\n",
       "0                       0.0000                       0.0000   \n",
       "1                       5.5624                       0.0000   \n",
       "2                      14.3711                       6.3091   \n",
       "3                     128.3151                      53.6278   \n",
       "4                     697.6744                    1245.2681   \n",
       "\n",
       "  TCGA-WT-AB44-01A-11R-A41B-07 TCGA-XX-A899-01A-11R-A36F-07  \\\n",
       "0                       0.0000                       0.0000   \n",
       "1                       0.0000                      14.3858   \n",
       "2                       3.2580                      21.4409   \n",
       "3                      42.2643                     137.7756   \n",
       "4                    1877.4180                     652.7559   \n",
       "\n",
       "  TCGA-XX-A89A-01A-11R-A36F-07 TCGA-Z7-A8R5-01A-42R-A41B-07  \\\n",
       "0                       0.0000                       0.0000   \n",
       "1                      22.3240                       2.2638   \n",
       "2                      27.2744                       7.2933   \n",
       "3                      64.1427                      85.0461   \n",
       "4                     722.7208                    1140.2801   \n",
       "\n",
       "  TCGA-Z7-A8R6-01A-11R-A41B-07  \n",
       "0                       0.0000  \n",
       "1                       6.8865  \n",
       "2                      24.7795  \n",
       "3                     167.5511  \n",
       "4                    1003.5668  \n",
       "\n",
       "[5 rows x 1095 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genome_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_genome_new.to_csv(\"genome_data.csv\",sep=\",\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Creating new clinical DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Adding a new column with the references only (TCGA-XXXX-XX format)\n",
    "df_clin[\"reference\"]= [val[0:12] for val in df_clin[\"submitter_id_demo\"].astype(str).tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Creating new clinical dataframe\n",
    "to_remove=[\"case_id\",\"demographic_id\",\"submitter_id_demo\",\"updated_datetime_demo\",\n",
    "          \"classification_of_tumor\",\"days_to_last_follow_up\",\"diagnosis_id\",\"last_known_disease_status\",\n",
    "          \"prior_malignancy\",\"progression_or_recurrence\",\"tumor_grade\",\"exposure_id\",\"submitter_id\",\n",
    "          \"updated_datetime_diag\", \"updated_datetime\",\"submitter_id_diag\",\"primary_diagnosis\",\n",
    "           \"site_of_resection_or_biopsy\",\"days_to_birth\",\"year_of_death\"]\n",
    "df_clin_new=df_clin.drop(to_remove,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_clin_new=df_clin_new[df_clin_new[\"reference\"].isin([val[0:12] for val in df_genome.columns.tolist()[1:] if val[13:15]==\"01\"])]\n",
    "df_clin_new.reset_index(drop=True,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Fixing year of birth\n",
    "df_clin_new[\"year_of_birth\"]=[int(x) if not math.isnan(x) else x for x in df_clin_new.year_of_birth.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Fixing age_at_diagnosis\n",
    "df_clin_new[\"age_at_diagnosis\"]=df_clin_new.age_at_diagnosis.apply(lambda x: round(float(x)/365,0) if not math.isnan(float(x)) else None)\n",
    "df_clin_new[\"age_at_diagnosis\"]=[int(x) if not math.isnan(x) else x for x in df_clin_new.age_at_diagnosis.tolist()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": true,
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "#Fixing days_to_death\n",
    "df_clin_new[\"days_to_death\"]=df_clin_new.days_to_death.apply(lambda x: round(x/365,0) if not math.isnan(x) else None)\n",
    "df_clin_new[\"days_to_death\"]=[int(x) if not math.isnan(x) else x for x in df_clin_new.days_to_death.tolist()]\n",
    "df_clin_new.rename(columns={\"days_to_death\":\"years_to_death\"},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Mapping the morphology code to its name\n",
    "dict_morphology={\"8520/3\":\"Lobular Carcinoma\", #Second most common\n",
    "                 \"8500/3\":\"Infiltrating Duct Carcinoma\", #Most common\n",
    "                \"8575/3\":\"Metaplastic Carcinoma\",#Only in 1% of all breast cancer\n",
    "                \"8523/3\":\"Infiltrating Duct Mixed with Other Types Of Carcinoma\",\n",
    "                \"8211/3\":\"Tubular Adenocarcinoma\",\n",
    "                \"8522/3\":\"Infiltrating Duct and Lobular Carcinoma\",\n",
    "                \"9020/3\":\"Phyllodes Tumor\",\n",
    "                \"8541/3\":\"Paget Disease and Infiltrating Duct Carcinoma of Breast\",\n",
    "                \"8480/3\":\"Mucinous Adenocarcinoma\",\n",
    "                \"8503/3\":\"Intraductal Papillary Adenocarcinoma with Invasion\",\n",
    "                \"8090/3\":\"Basal Cell Carcinoma\",\n",
    "                \"8502/3\":\"Secretory Carcinoma of Breast\",\n",
    "                \"8510/3\":\"Medullary Carcinoma\",\n",
    "                \"8524/3\":\"Infiltrating Lobular Mixed with Other Types of Carcinoma\",\n",
    "                \"8010/3\":\"Carcinoma\",\n",
    "                \"8507/3\":\"Infiltrating Ductal\",\n",
    "                \"8022/3\":\"Pleomorphic Carcinoma\",\n",
    "                \"8050/3\":\"Papillary Carcinoma\",\n",
    "                \"8201/3\":\"Cribriform Carcinoma\",\n",
    "                \"8013/3\":\"Large Cell Neuroendocrine Carcinoma\",\n",
    "                \"8200/3\":\"Adenoid Cystic Carcinoma\",\n",
    "                \"8401/3\":\"Apocrine Adenocarcinoma\"} \n",
    "df_clin_new[\"morphology_code\"]=df_clin_new[\"morphology\"].tolist()\n",
    "df_clin_new.replace({\"morphology\":dict_morphology},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Mapping the tissue or organ of origin\n",
    "dict_tissue={\"c50.9\":\"Malignant neoplasm of breast of unspecified site\",\n",
    "            \"c50.8\":\"Malignant neoplasm of overlapping sites of breast\",\n",
    "            \"c50.3\":\"Malignant neoplasm of lower-inner quadrant of breast\",\n",
    "            \"c50.4\":\"Malignant neoplasm of upper-outer quadrant of breast\",\n",
    "            \"c50.2\":\"Malignant neoplasm of upper-inner quadrant of breast\",\n",
    "            \"c50.5\":\"Malignant neoplasm of lower-outer quadrant of breast\"}\n",
    "df_clin_new[\"tissue_of_origin_code\"]=df_clin_new[\"tissue_or_organ_of_origin\"].tolist()\n",
    "df_clin_new.replace({\"tissue_or_organ_of_origin\":dict_tissue},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Not considering normal tissue: Getting the center code\n",
    "list_ref=[val for val in df_genome.columns.tolist()[1:] if (val[13:15]==\"01\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Characterization_center\n",
    "dict_center={\"07\":\"University of North Carolina\"}\n",
    "dict_center_code={}\n",
    "for patient in list_ref:\n",
    "    dict_center_code[patient[0:12]]=patient[26:28]\n",
    "df_clin_new[\"characterization_center_code\"]=df_clin_new[\"reference\"].tolist()\n",
    "df_clin_new.replace({\"characterization_center_code\":dict_center_code},inplace=True)\n",
    "df_clin_new[\"characterization_center\"]=df_clin_new[\"characterization_center_code\"].tolist()\n",
    "df_clin_new.replace({\"characterization_center\":dict_center},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Getting the tissue code\n",
    "dict_tissue_code={}\n",
    "for patient in list_ref:\n",
    "    dict_tissue_code[patient[0:12]]=patient[5:7]\n",
    "df_clin_new[\"tissue_source_code\"]=df_clin_new[\"reference\"].tolist()\n",
    "df_clin_new.replace({\"tissue_source_code\":dict_tissue_code},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Creating the dictionary for the tissue code and its values (Institute, Study name, Country and state)\n",
    "dict_tissue_source={\"A2\":[\"Walter Reed Army Medical Center\",\"Breast invasive carcinoma\",\"USA\",\"Washington D.C.\"],\n",
    "                         \"D8\":[\"Greater Poland Cancer Center\",\"Breast invasive carcinoma\",\"Poland\",\"Greater Poland\"],\n",
    "                         \"BH\":[\"University of Pittsburgh\",\"Breast invasive carcinoma\",\"USA\",\"Pennsylvania\"],\n",
    "                         \"E2\":[\"Roswell Park Cancer Institute\",\"Breast invasive carcinoma\",\"USA\",\"New York\"],\n",
    "                         \"C8\":[\"ILSBio\",\"Breast invasive carcinoma\",\"USA\",\"Maryland\"],\n",
    "                         \"AQ\":[\"University of North Carolina\",\"Breast invasive carcinoma\",\"USA\",\"North Carolina\"],\n",
    "                         \"AN\":[\"Cureline\",\"Breast invasive carcinoma\",\"USA\",\"California\"],\n",
    "                         \"V7\":[\"Medical College of Georgia\",\"Breast invasive carcinoma\",\"USA\",\"Georgia\"],\n",
    "                         \"S3\":[\"Albert Einstein Medical Center\",\"Breast invasive carcinoma\",\"USA\",\"Pennsylvania\"],\n",
    "                         \"A8\":[\"Indivumed\",\"Breast invasive carcinoma\",\"Germany\",\"Hamburg\"],\n",
    "                         \"GM\":[\"MD Anderson Cancer Center\",\"Breast invasive carcinoma\",\"USA\",\"Texas\"],\n",
    "                         \"AR\":[\"Mayo Clinic\",\"Breast invasive carcinoma\",\"USA\",\"Minnesota\"],\n",
    "                         \"E9\":[\"Asterand Bioscience\",\"Breast invasive carcinoma\",\"USA\",\"Michigan\"],\n",
    "                         \"A7\":[\"Christiana Care Health System\",\"Breast invasive carcinoma\",\"USA\",\"Delaware\"],\n",
    "                         \"AO\":[\"Memorial Sloan Kettering Cancer Center\",\"Breast invasive carcinoma\",\"USA\",\"NY\"],\n",
    "                         \"B6\":[\"Duke Cancer Institute\",\"Breast invasive carcinoma\",\"USA\",\"North Carolina\"],\n",
    "                         \"AC\":[\"International Genomics Consortium\",\"Breast invasive carcinoma\",\"USA\",\"Arizona\"],\n",
    "                         \"PE\":[\"Fox Chase Cancer Center\",\"Breast invasive carcinoma\",\"USA\",\"Pennsylvania\"],\n",
    "                         \"EW\":[\"University of Miami\",\"Breast invasive carcinoma\",\"USA\",\"Florida\"],\n",
    "                         \"XX\":[\"Spectrum Health\",\"Breast invasive carcinoma\",\"USA\",\"Michigan\"],\n",
    "                         \"OL\":[\"University of Chicago\",\"Breast invasive carcinoma\",\"USA\",\"Illinois\"],\n",
    "                         \"GI\":[\"Indiana University–Purdue University Indianapolis\",\"Breast invasive carcinoma\",\"USA\",\"Indiana\"],\n",
    "                         \"3C\":[\"Columbia University\",\"Breast invasive carcinoma\",\"USA\",\"New York\"],\n",
    "                         \"LL\":[\"Candler Hospital\",\"Breast invasive carcinoma\",\"USA\",\"Georgia\"],\n",
    "                         \"OK\":[\"Mount Sinai School of Medicine\",\"Breast invasive carcinoma\",\"USA\",\"New York\"],\n",
    "                         \"WT\":[\"University of Kansas\",\"Breast invasive carcinoma\",\"USA\",\"Kansas\"],\n",
    "                         \"JL\":[\"Metrics Research Pakistan\",\"Breast invasive carcinoma\",\"Pakistan\",\"Sindh\"],\n",
    "                         \"LD\":[\"Hartford Hospital\",\"Breast invasive carcinoma\",\"USA\",\"Connecticut\"],\n",
    "                         \"UL\":[\"Boston Medical Center\",\"Breast invasive carcinoma\",\"USA\",\"California\"],\n",
    "                         \"A1\":[\"University of California San Francisco\",\"Breast invasive carcinoma\",\"USA\",\"California\"],\n",
    "                         \"PL\":[\"Institute of Human Virology Nigeria\",\"Breast invasive carcinoma\",\"Nigeria\",\"Federal Capital Territory\"],\n",
    "                         \"HN\":[\"Ontario Institute for Cancer Research\",\"Breast invasive carcinoma\",\"Canada\",\"Ontario\"],\n",
    "                         \"5L\":[\"University of Sao Paulo\",\"Breast invasive carcinoma\",\"Brazil\",\"Sao Paulo\"],\n",
    "                         \"Z7\":[\"John Wayne Cancer Institute\",\"Breast invasive carcinoma\",\"USA\",\"California\"],\n",
    "                         \"5T\":[\"Holy Cross Health Center\",\"Breast invasive carcinoma\",\"USA\",\"Maryland\"],\n",
    "                         \"UU\":[\"Mary Bird Perkins Cancer Center - Our Lady of the Lake\",\"Breast invasive carcinoma\",\"USA\",\"Louisiana\"],\n",
    "                         \"LQ\":[\"Gundersen Lutheran Health System\",\"Breast invasive carcinoma\",\"USA\",\"Wisconsin\"],\n",
    "                         \"4H\":[\"Proteogenex Inc\",\"Breast invasive carcinoma\",\"USA\",\"California\"],\n",
    "                         \"MS\":[\"University of Minnesota\",\"Breast invasive carcinoma\",\"USA\",\"Minnesota\"],\n",
    "                         \"W8\":[\"Greenville Health System\",\"Breast invasive carcinoma\",\"USA\",\"South Carolina\"]\n",
    "                        }\n",
    "dict_tissue_source_institute={}\n",
    "dict_tissue_source_study_name={}\n",
    "dict_tissue_source_country={}\n",
    "dict_tissue_source_state={}\n",
    "for code in dict_tissue_source:\n",
    "    dict_tissue_source_institute[code]=dict_tissue_source[code][0]\n",
    "    dict_tissue_source_study_name[code]=dict_tissue_source[code][1]\n",
    "    dict_tissue_source_country[code]=dict_tissue_source[code][2]\n",
    "    dict_tissue_source_state[code]=dict_tissue_source[code][3]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Mapping tissue codes to its values\n",
    "df_clin_new[\"tissue_source_institute\"]=df_clin_new[\"tissue_source_code\"].tolist()\n",
    "df_clin_new[\"tissue_source_study_name\"]=df_clin_new[\"tissue_source_code\"].tolist()\n",
    "df_clin_new[\"tissue_source_country\"]=df_clin_new[\"tissue_source_code\"].tolist()\n",
    "df_clin_new[\"tissue_source_state\"]=df_clin_new[\"tissue_source_code\"].tolist()\n",
    "df_clin_new.replace({\"tissue_source_institute\":dict_tissue_source_institute},inplace=True)\n",
    "df_clin_new.replace({\"tissue_source_study_name\":dict_tissue_source_study_name},inplace=True)\n",
    "df_clin_new.replace({\"tissue_source_country\":dict_tissue_source_country},inplace=True)\n",
    "df_clin_new.replace({\"tissue_source_state\":dict_tissue_source_state},inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Rearranging the columns\n",
    "ordered_cols=[\"reference\",\"gender\",\"ethnicity\",\"race\",\"vital_status\",\"year_of_birth\",\"age_at_diagnosis\",\n",
    "              \"years_to_death\",\"tissue_of_origin_code\",\"tissue_or_organ_of_origin\",\"morphology_code\",\"morphology\",\n",
    "              \"tumor_stage\",\"tissue_source_code\",\"tissue_source_institute\",\"tissue_source_study_name\",\n",
    "              \"tissue_source_country\",\"tissue_source_state\",\"characterization_center_code\",\"characterization_center\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_clin_new=df_clin_new[ordered_cols]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Cleaning races names\n",
    "#Replacing latino in the race cells that are marked as not-reported and which ethnicity is latino.\n",
    "for n,race in enumerate(df_clin_new.race.tolist()):\n",
    "    if df_clin_new.race[n]==\"not reported\" and df_clin_new.ethnicity[n]==\"hispanic or latino\":\n",
    "        df_clin_new.at[n,\"race\"]=\"latino\"\n",
    "dict_races={\"black or african american\":\"black\",\"american indian or alaska native\":\"native american\"}\n",
    "df_clin_new.replace({\"race\":dict_races},inplace=True)\n",
    "\n",
    "#Dropping ethnicity column\n",
    "df_clin_new.drop([\"ethnicity\"],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>year_of_birth</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>years_to_death</th>\n",
       "      <th>tissue_of_origin_code</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>morphology_code</th>\n",
       "      <th>morphology</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>tissue_source_code</th>\n",
       "      <th>tissue_source_institute</th>\n",
       "      <th>tissue_source_study_name</th>\n",
       "      <th>tissue_source_country</th>\n",
       "      <th>tissue_source_state</th>\n",
       "      <th>characterization_center_code</th>\n",
       "      <th>characterization_center</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A2-A0YL</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>alive</td>\n",
       "      <td>1962.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>Malignant neoplasm of breast of unspecified site</td>\n",
       "      <td>8520/3</td>\n",
       "      <td>Lobular Carcinoma</td>\n",
       "      <td>stage iiia</td>\n",
       "      <td>A2</td>\n",
       "      <td>Walter Reed Army Medical Center</td>\n",
       "      <td>Breast invasive carcinoma</td>\n",
       "      <td>USA</td>\n",
       "      <td>Washington D.C.</td>\n",
       "      <td>07</td>\n",
       "      <td>University of North Carolina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-D8-A1XR</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>alive</td>\n",
       "      <td>1954.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>Malignant neoplasm of breast of unspecified site</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating Duct Carcinoma</td>\n",
       "      <td>stage iib</td>\n",
       "      <td>D8</td>\n",
       "      <td>Greater Poland Cancer Center</td>\n",
       "      <td>Breast invasive carcinoma</td>\n",
       "      <td>Poland</td>\n",
       "      <td>Greater Poland</td>\n",
       "      <td>07</td>\n",
       "      <td>University of North Carolina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-BH-A1ET</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>dead</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>Malignant neoplasm of breast of unspecified site</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating Duct Carcinoma</td>\n",
       "      <td>stage i</td>\n",
       "      <td>BH</td>\n",
       "      <td>University of Pittsburgh</td>\n",
       "      <td>Breast invasive carcinoma</td>\n",
       "      <td>USA</td>\n",
       "      <td>Pennsylvania</td>\n",
       "      <td>07</td>\n",
       "      <td>University of North Carolina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-E2-A1LE</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>dead</td>\n",
       "      <td>1937.0</td>\n",
       "      <td>72.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>Malignant neoplasm of breast of unspecified site</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating Duct Carcinoma</td>\n",
       "      <td>stage iiic</td>\n",
       "      <td>E2</td>\n",
       "      <td>Roswell Park Cancer Institute</td>\n",
       "      <td>Breast invasive carcinoma</td>\n",
       "      <td>USA</td>\n",
       "      <td>New York</td>\n",
       "      <td>07</td>\n",
       "      <td>University of North Carolina</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-C8-A12V</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>alive</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>56.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>Malignant neoplasm of breast of unspecified site</td>\n",
       "      <td>8500/3</td>\n",
       "      <td>Infiltrating Duct Carcinoma</td>\n",
       "      <td>stage iia</td>\n",
       "      <td>C8</td>\n",
       "      <td>ILSBio</td>\n",
       "      <td>Breast invasive carcinoma</td>\n",
       "      <td>USA</td>\n",
       "      <td>Maryland</td>\n",
       "      <td>07</td>\n",
       "      <td>University of North Carolina</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      reference  gender   race vital_status  year_of_birth  age_at_diagnosis  \\\n",
       "0  TCGA-A2-A0YL  female  white        alive         1962.0              48.0   \n",
       "1  TCGA-D8-A1XR  female  white        alive         1954.0              56.0   \n",
       "2  TCGA-BH-A1ET  female  white         dead         1944.0              56.0   \n",
       "3  TCGA-E2-A1LE  female  white         dead         1937.0              72.0   \n",
       "4  TCGA-C8-A12V  female  asian        alive         1955.0              56.0   \n",
       "\n",
       "   years_to_death tissue_of_origin_code  \\\n",
       "0             NaN                 c50.9   \n",
       "1             NaN                 c50.9   \n",
       "2             7.0                 c50.9   \n",
       "3             2.0                 c50.9   \n",
       "4             NaN                 c50.9   \n",
       "\n",
       "                          tissue_or_organ_of_origin morphology_code  \\\n",
       "0  Malignant neoplasm of breast of unspecified site          8520/3   \n",
       "1  Malignant neoplasm of breast of unspecified site          8500/3   \n",
       "2  Malignant neoplasm of breast of unspecified site          8500/3   \n",
       "3  Malignant neoplasm of breast of unspecified site          8500/3   \n",
       "4  Malignant neoplasm of breast of unspecified site          8500/3   \n",
       "\n",
       "                    morphology tumor_stage tissue_source_code  \\\n",
       "0            Lobular Carcinoma  stage iiia                 A2   \n",
       "1  Infiltrating Duct Carcinoma   stage iib                 D8   \n",
       "2  Infiltrating Duct Carcinoma     stage i                 BH   \n",
       "3  Infiltrating Duct Carcinoma  stage iiic                 E2   \n",
       "4  Infiltrating Duct Carcinoma   stage iia                 C8   \n",
       "\n",
       "           tissue_source_institute   tissue_source_study_name  \\\n",
       "0  Walter Reed Army Medical Center  Breast invasive carcinoma   \n",
       "1     Greater Poland Cancer Center  Breast invasive carcinoma   \n",
       "2         University of Pittsburgh  Breast invasive carcinoma   \n",
       "3    Roswell Park Cancer Institute  Breast invasive carcinoma   \n",
       "4                           ILSBio  Breast invasive carcinoma   \n",
       "\n",
       "  tissue_source_country tissue_source_state characterization_center_code  \\\n",
       "0                   USA     Washington D.C.                           07   \n",
       "1                Poland      Greater Poland                           07   \n",
       "2                   USA        Pennsylvania                           07   \n",
       "3                   USA            New York                           07   \n",
       "4                   USA            Maryland                           07   \n",
       "\n",
       "        characterization_center  \n",
       "0  University of North Carolina  \n",
       "1  University of North Carolina  \n",
       "2  University of North Carolina  \n",
       "3  University of North Carolina  \n",
       "4  University of North Carolina  "
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clin_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Since those are too many headers to display, we will only focus on the following: \n",
    "#references, gender, race, age at diagnosis and tissue source country\n",
    "df_clin_viz=df_clin_new[[\"reference\",\"gender\",\"race\",\"age_at_diagnosis\",\"tissue_source_country\"]].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "groups=[]\n",
    "for row in df_clin_viz[\"age_at_diagnosis\"]:\n",
    "    if row<50:\n",
    "        groups.append(\"less_than_50\")\n",
    "    if row>=50 and row <60:\n",
    "        groups.append(\"between_50_and_60\")\n",
    "    if row>=60 and row <70:\n",
    "        groups.append(\"between_60_and_70\")\n",
    "    if row>=70:\n",
    "        groups.append(\"more_than_70\")\n",
    "    if math.isnan(row):\n",
    "        groups.append(\"not_reported\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_clin_viz[\"age_groups\"]=groups\n",
    "df_clin_viz.drop([\"age_at_diagnosis\"],axis=1,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>tissue_source_country</th>\n",
       "      <th>age_groups</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-A2-A0YL</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>USA</td>\n",
       "      <td>less_than_50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-D8-A1XR</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>Poland</td>\n",
       "      <td>between_50_and_60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-BH-A1ET</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>USA</td>\n",
       "      <td>between_50_and_60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-E2-A1LE</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>USA</td>\n",
       "      <td>more_than_70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-C8-A12V</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>USA</td>\n",
       "      <td>between_50_and_60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      reference  gender   race tissue_source_country         age_groups\n",
       "0  TCGA-A2-A0YL  female  white                   USA       less_than_50\n",
       "1  TCGA-D8-A1XR  female  white                Poland  between_50_and_60\n",
       "2  TCGA-BH-A1ET  female  white                   USA  between_50_and_60\n",
       "3  TCGA-E2-A1LE  female  white                   USA       more_than_70\n",
       "4  TCGA-C8-A12V  female  asian                   USA  between_50_and_60"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clin_viz.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>reference</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>tissue_source_country</th>\n",
       "      <th>age_groups</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"36\" valign=\"top\">female</th>\n",
       "      <th rowspan=\"6\" valign=\"top\">asian</th>\n",
       "      <th>Pakistan</th>\n",
       "      <th>not_reported</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">USA</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>more_than_70</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>not_reported</th>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"7\" valign=\"top\">black</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Nigeria</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">USA</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>53</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>more_than_70</th>\n",
       "      <td>37</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>not_reported</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">latino</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">USA</th>\n",
       "      <th>less_than_50</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>more_than_70</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>native american</th>\n",
       "      <th>USA</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"9\" valign=\"top\">not reported</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Canada</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Germany</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>more_than_70</th>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">USA</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"11\" valign=\"top\">white</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">Brazil</th>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">Poland</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>more_than_70</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"5\" valign=\"top\">USA</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>164</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>193</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>less_than_50</th>\n",
       "      <td>177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>more_than_70</th>\n",
       "      <td>131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>not_reported</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"6\" valign=\"top\">male</th>\n",
       "      <th>black</th>\n",
       "      <th>USA</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>not reported</th>\n",
       "      <th>Germany</th>\n",
       "      <th>less_than_50</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"4\" valign=\"top\">white</th>\n",
       "      <th>Poland</th>\n",
       "      <th>less_than_50</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">USA</th>\n",
       "      <th>between_50_and_60</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>between_60_and_70</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>more_than_70</th>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                reference\n",
       "gender race            tissue_source_country age_groups                  \n",
       "female asian           Pakistan              not_reported               2\n",
       "                       USA                   between_50_and_60         19\n",
       "                                             between_60_and_70          9\n",
       "                                             less_than_50              14\n",
       "                                             more_than_70               6\n",
       "                                             not_reported              11\n",
       "       black           Nigeria               between_50_and_60          1\n",
       "                                             less_than_50               3\n",
       "                       USA                   between_50_and_60         48\n",
       "                                             between_60_and_70         36\n",
       "                                             less_than_50              53\n",
       "                                             more_than_70              37\n",
       "                                             not_reported               2\n",
       "       latino          USA                   less_than_50               1\n",
       "                                             more_than_70               1\n",
       "       native american USA                   between_50_and_60          1\n",
       "       not reported    Canada                between_50_and_60          1\n",
       "                                             less_than_50               1\n",
       "                       Germany               between_50_and_60         14\n",
       "                                             between_60_and_70         26\n",
       "                                             less_than_50               9\n",
       "                                             more_than_70              34\n",
       "                       USA                   between_50_and_60          1\n",
       "                                             between_60_and_70          4\n",
       "                                             less_than_50               2\n",
       "       white           Brazil                between_60_and_70          1\n",
       "                                             less_than_50               1\n",
       "                       Poland                between_50_and_60         22\n",
       "                                             between_60_and_70         18\n",
       "                                             less_than_50              13\n",
       "                                             more_than_70              24\n",
       "                       USA                   between_50_and_60        164\n",
       "                                             between_60_and_70        193\n",
       "                                             less_than_50             177\n",
       "                                             more_than_70             131\n",
       "                                             not_reported               1\n",
       "male   black           USA                   between_50_and_60          3\n",
       "       not reported    Germany               less_than_50               1\n",
       "       white           Poland                less_than_50               1\n",
       "                       USA                   between_50_and_60          1\n",
       "                                             between_60_and_70          3\n",
       "                                             more_than_70               3"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clin_viz.groupby([\"gender\",\"race\",\"tissue_source_country\",\"age_groups\"]).count()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_clin_viz.to_csv(\"clinical_data.csv\",sep=\",\",index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## New distances matrix DataFrame"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_distances_new = df_distances.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "columns_dict = {}\n",
    "for n,key in enumerate(df_distances.columns.tolist()):\n",
    "    columns_dict[key] = [val.replace(\".\",\"-\")[0:12] for val in df_distances.columns.tolist()][n]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Just to keep the same format, we rename the distances matrix headers and indices\n",
    "df_distances_new.rename(columns=columns_dict,index=columns_dict,inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-3C-AAAU</th>\n",
       "      <th>TCGA-3C-AALI</th>\n",
       "      <th>TCGA-3C-AALJ</th>\n",
       "      <th>TCGA-3C-AALK</th>\n",
       "      <th>TCGA-4H-AAAK</th>\n",
       "      <th>TCGA-5L-AAT0</th>\n",
       "      <th>TCGA-5L-AAT1</th>\n",
       "      <th>TCGA-5T-A9QA</th>\n",
       "      <th>TCGA-A1-A0SB</th>\n",
       "      <th>TCGA-A1-A0SD</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-UL-AAZ6</th>\n",
       "      <th>TCGA-UU-A93S</th>\n",
       "      <th>TCGA-V7-A7HQ</th>\n",
       "      <th>TCGA-W8-A86G</th>\n",
       "      <th>TCGA-WT-AB41</th>\n",
       "      <th>TCGA-WT-AB44</th>\n",
       "      <th>TCGA-XX-A899</th>\n",
       "      <th>TCGA-XX-A89A</th>\n",
       "      <th>TCGA-Z7-A8R5</th>\n",
       "      <th>TCGA-Z7-A8R6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AAAU</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.120880</td>\n",
       "      <td>0.105801</td>\n",
       "      <td>0.109540</td>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.116020</td>\n",
       "      <td>0.138326</td>\n",
       "      <td>0.139012</td>\n",
       "      <td>0.164870</td>\n",
       "      <td>0.109789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130849</td>\n",
       "      <td>0.177643</td>\n",
       "      <td>0.202318</td>\n",
       "      <td>0.118387</td>\n",
       "      <td>0.204810</td>\n",
       "      <td>0.172970</td>\n",
       "      <td>0.131472</td>\n",
       "      <td>0.147361</td>\n",
       "      <td>0.160945</td>\n",
       "      <td>0.102312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALI</th>\n",
       "      <td>0.120880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.105240</td>\n",
       "      <td>0.102997</td>\n",
       "      <td>0.095894</td>\n",
       "      <td>0.142688</td>\n",
       "      <td>0.183251</td>\n",
       "      <td>0.114400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117577</td>\n",
       "      <td>0.143124</td>\n",
       "      <td>0.172721</td>\n",
       "      <td>0.125615</td>\n",
       "      <td>0.158141</td>\n",
       "      <td>0.134712</td>\n",
       "      <td>0.116518</td>\n",
       "      <td>0.119447</td>\n",
       "      <td>0.129105</td>\n",
       "      <td>0.101502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALJ</th>\n",
       "      <td>0.105801</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.072653</td>\n",
       "      <td>0.091719</td>\n",
       "      <td>0.082123</td>\n",
       "      <td>0.096455</td>\n",
       "      <td>0.141006</td>\n",
       "      <td>0.181756</td>\n",
       "      <td>0.102686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108979</td>\n",
       "      <td>0.147735</td>\n",
       "      <td>0.151162</td>\n",
       "      <td>0.113216</td>\n",
       "      <td>0.138139</td>\n",
       "      <td>0.120631</td>\n",
       "      <td>0.100941</td>\n",
       "      <td>0.103558</td>\n",
       "      <td>0.112343</td>\n",
       "      <td>0.083557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-3C-AALK</th>\n",
       "      <td>0.109540</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.072653</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040750</td>\n",
       "      <td>0.040376</td>\n",
       "      <td>0.062932</td>\n",
       "      <td>0.144121</td>\n",
       "      <td>0.129292</td>\n",
       "      <td>0.053960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117827</td>\n",
       "      <td>0.158452</td>\n",
       "      <td>0.146738</td>\n",
       "      <td>0.062870</td>\n",
       "      <td>0.165742</td>\n",
       "      <td>0.102312</td>\n",
       "      <td>0.060814</td>\n",
       "      <td>0.073400</td>\n",
       "      <td>0.080130</td>\n",
       "      <td>0.070721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA-4H-AAAK</th>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.105240</td>\n",
       "      <td>0.091719</td>\n",
       "      <td>0.040750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037074</td>\n",
       "      <td>0.065612</td>\n",
       "      <td>0.152471</td>\n",
       "      <td>0.113901</td>\n",
       "      <td>0.054209</td>\n",
       "      <td>...</td>\n",
       "      <td>0.143560</td>\n",
       "      <td>0.180884</td>\n",
       "      <td>0.172347</td>\n",
       "      <td>0.065300</td>\n",
       "      <td>0.187800</td>\n",
       "      <td>0.112343</td>\n",
       "      <td>0.059381</td>\n",
       "      <td>0.069163</td>\n",
       "      <td>0.097015</td>\n",
       "      <td>0.078198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1093 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "              TCGA-3C-AAAU  TCGA-3C-AALI  TCGA-3C-AALJ  TCGA-3C-AALK  \\\n",
       "TCGA-3C-AAAU      0.000000      0.120880      0.105801      0.109540   \n",
       "TCGA-3C-AALI      0.120880      0.000000      0.086734      0.086734   \n",
       "TCGA-3C-AALJ      0.105801      0.086734      0.000000      0.072653   \n",
       "TCGA-3C-AALK      0.109540      0.086734      0.072653      0.000000   \n",
       "TCGA-4H-AAAK      0.118014      0.105240      0.091719      0.040750   \n",
       "\n",
       "              TCGA-4H-AAAK  TCGA-5L-AAT0  TCGA-5L-AAT1  TCGA-5T-A9QA  \\\n",
       "TCGA-3C-AAAU      0.118014      0.116020      0.138326      0.139012   \n",
       "TCGA-3C-AALI      0.105240      0.102997      0.095894      0.142688   \n",
       "TCGA-3C-AALJ      0.091719      0.082123      0.096455      0.141006   \n",
       "TCGA-3C-AALK      0.040750      0.040376      0.062932      0.144121   \n",
       "TCGA-4H-AAAK      0.000000      0.037074      0.065612      0.152471   \n",
       "\n",
       "              TCGA-A1-A0SB  TCGA-A1-A0SD      ...       TCGA-UL-AAZ6  \\\n",
       "TCGA-3C-AAAU      0.164870      0.109789      ...           0.130849   \n",
       "TCGA-3C-AALI      0.183251      0.114400      ...           0.117577   \n",
       "TCGA-3C-AALJ      0.181756      0.102686      ...           0.108979   \n",
       "TCGA-3C-AALK      0.129292      0.053960      ...           0.117827   \n",
       "TCGA-4H-AAAK      0.113901      0.054209      ...           0.143560   \n",
       "\n",
       "              TCGA-UU-A93S  TCGA-V7-A7HQ  TCGA-W8-A86G  TCGA-WT-AB41  \\\n",
       "TCGA-3C-AAAU      0.177643      0.202318      0.118387      0.204810   \n",
       "TCGA-3C-AALI      0.143124      0.172721      0.125615      0.158141   \n",
       "TCGA-3C-AALJ      0.147735      0.151162      0.113216      0.138139   \n",
       "TCGA-3C-AALK      0.158452      0.146738      0.062870      0.165742   \n",
       "TCGA-4H-AAAK      0.180884      0.172347      0.065300      0.187800   \n",
       "\n",
       "              TCGA-WT-AB44  TCGA-XX-A899  TCGA-XX-A89A  TCGA-Z7-A8R5  \\\n",
       "TCGA-3C-AAAU      0.172970      0.131472      0.147361      0.160945   \n",
       "TCGA-3C-AALI      0.134712      0.116518      0.119447      0.129105   \n",
       "TCGA-3C-AALJ      0.120631      0.100941      0.103558      0.112343   \n",
       "TCGA-3C-AALK      0.102312      0.060814      0.073400      0.080130   \n",
       "TCGA-4H-AAAK      0.112343      0.059381      0.069163      0.097015   \n",
       "\n",
       "              TCGA-Z7-A8R6  \n",
       "TCGA-3C-AAAU      0.102312  \n",
       "TCGA-3C-AALI      0.101502  \n",
       "TCGA-3C-AALJ      0.083557  \n",
       "TCGA-3C-AALK      0.070721  \n",
       "TCGA-4H-AAAK      0.078198  \n",
       "\n",
       "[5 rows x 1093 columns]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_distances_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "df_distances_new.to_csv(\"distances_matrix.csv\",sep=\",\",index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
