{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import json"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Directories of the TCGA data\n",
    "CLIN_PATH = '../data/clinical.project-TCGA-BRCA.2016-09-30T12_32_20.666404.json' #Patient data\n",
    "DIST_PATH = '../data/dd5_5_for_TCGA.txt' # Distance metric\n",
    "GENOME_PATH = '../data/BRCA.rnaseqv2__illuminahiseq_rnaseqv2__unc_edu__Level_3__RSEM_genes_normalized__data.data.txt' #Genome data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Clinical data of patients"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clinical data of patients\n",
    "big_df = pd.read_json(CLIN_PATH)\n",
    "\n",
    "# Change lists to dictionaries to easily expand to columns in DataFrame\n",
    "def dictify(df):\n",
    "    df['diagnoses'] = df['diagnoses'][0]\n",
    "    df['exposures'] = df['exposures'][0]\n",
    "    return df\n",
    "\n",
    "big_df = big_df.dropna().apply(dictify,axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Expand dictionaries to columns in DataFrame\n",
    "df_clin = pd.DataFrame(big_df['case_id']).join(big_df['demographic'].apply(pd.Series),how='outer',lsuffix='_case_id',rsuffix='_demo')\\\n",
    "            .join(big_df['diagnoses'].apply(pd.Series),how='outer',lsuffix = '_demo',rsuffix='_diag')\\\n",
    "            .join(big_df['exposures'].apply(pd.Series),how='outer',lsuffix = '_diag',rsuffix='_expo')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# Drop nulls rows and columns\n",
    "df_clin.dropna(how='all', inplace = True)\n",
    "df_clin.dropna(how='all', inplace = True, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>case_id</th>\n",
       "      <th>demographic_id</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>race</th>\n",
       "      <th>submitter_id_demo</th>\n",
       "      <th>updated_datetime_demo</th>\n",
       "      <th>year_of_birth</th>\n",
       "      <th>year_of_death</th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>...</th>\n",
       "      <th>site_of_resection_or_biopsy</th>\n",
       "      <th>submitter_id_diag</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>tumor_stage</th>\n",
       "      <th>updated_datetime_diag</th>\n",
       "      <th>vital_status</th>\n",
       "      <th>exposure_id</th>\n",
       "      <th>submitter_id</th>\n",
       "      <th>updated_datetime</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b205c89f-af62-4186-acad-ed23d243fa98</td>\n",
       "      <td>55ae5a3b-e509-5fcd-98c9-c5007b448890</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-A2-A0YL_demographic</td>\n",
       "      <td>2016-09-02T19:00:43.039678-05:00</td>\n",
       "      <td>1962.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17702.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-A2-A0YL_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iiia</td>\n",
       "      <td>2016-09-02T19:00:43.039678-05:00</td>\n",
       "      <td>alive</td>\n",
       "      <td>23b6a296-e149-5801-aa20-e144ef683abe</td>\n",
       "      <td>TCGA-A2-A0YL_exposure</td>\n",
       "      <td>2016-09-02T19:00:43.039678-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>70f34c5c-3671-44c8-9469-99f9786efec1</td>\n",
       "      <td>498e27fb-2201-524a-bdd1-86f4e53dc54b</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-D8-A1XR_demographic</td>\n",
       "      <td>2016-09-02T19:07:03.114741-05:00</td>\n",
       "      <td>1954.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20488.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-D8-A1XR_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iib</td>\n",
       "      <td>2016-09-02T19:07:03.114741-05:00</td>\n",
       "      <td>alive</td>\n",
       "      <td>6dfa11dc-4266-57c9-b746-02a6c7bb9d23</td>\n",
       "      <td>TCGA-D8-A1XR_exposure</td>\n",
       "      <td>2016-09-02T19:07:03.114741-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8986a141-eae7-4157-b695-02cc6fc3b071</td>\n",
       "      <td>8faa57d4-1199-53ce-a0ce-24d358841153</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-BH-A1ET_demographic</td>\n",
       "      <td>2016-09-02T18:57:52.265307-05:00</td>\n",
       "      <td>1944.0</td>\n",
       "      <td>2005.0</td>\n",
       "      <td>20425.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-BH-A1ET_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage i</td>\n",
       "      <td>2016-09-02T18:57:52.265307-05:00</td>\n",
       "      <td>dead</td>\n",
       "      <td>e72e6b02-58be-539a-af7f-f2714f14899a</td>\n",
       "      <td>TCGA-BH-A1ET_exposure</td>\n",
       "      <td>2016-09-02T18:57:52.265307-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7317e605-93ba-49a9-8743-20613b02767a</td>\n",
       "      <td>4394b44c-b51d-5b96-b62b-db7f49007b43</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>white</td>\n",
       "      <td>TCGA-E2-A1LE_demographic</td>\n",
       "      <td>2016-09-02T19:09:17.163541-05:00</td>\n",
       "      <td>1937.0</td>\n",
       "      <td>2010.0</td>\n",
       "      <td>26239.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-E2-A1LE_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iiic</td>\n",
       "      <td>2016-09-02T19:09:17.163541-05:00</td>\n",
       "      <td>dead</td>\n",
       "      <td>4c456a08-ffda-507f-a3e3-d273e5c49864</td>\n",
       "      <td>TCGA-E2-A1LE_exposure</td>\n",
       "      <td>2016-09-02T19:09:17.163541-05:00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>7e1673f8-5758-4963-8804-d5e39f06205b</td>\n",
       "      <td>eebe9198-74c8-5996-bf88-c56d5fd6832f</td>\n",
       "      <td>not hispanic or latino</td>\n",
       "      <td>female</td>\n",
       "      <td>asian</td>\n",
       "      <td>TCGA-C8-A12V_demographic</td>\n",
       "      <td>2016-09-02T19:02:30.147198-05:00</td>\n",
       "      <td>1955.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20346.0</td>\n",
       "      <td>...</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>TCGA-C8-A12V_diagnosis</td>\n",
       "      <td>c50.9</td>\n",
       "      <td>not reported</td>\n",
       "      <td>stage iia</td>\n",
       "      <td>2016-09-02T19:02:30.147198-05:00</td>\n",
       "      <td>alive</td>\n",
       "      <td>dd5fbf53-64a4-5bc0-8ca7-ae05abbcbbac</td>\n",
       "      <td>TCGA-C8-A12V_exposure</td>\n",
       "      <td>2016-09-02T19:02:30.147198-05:00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 30 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                case_id                        demographic_id  \\\n",
       "0  b205c89f-af62-4186-acad-ed23d243fa98  55ae5a3b-e509-5fcd-98c9-c5007b448890   \n",
       "1  70f34c5c-3671-44c8-9469-99f9786efec1  498e27fb-2201-524a-bdd1-86f4e53dc54b   \n",
       "2  8986a141-eae7-4157-b695-02cc6fc3b071  8faa57d4-1199-53ce-a0ce-24d358841153   \n",
       "3  7317e605-93ba-49a9-8743-20613b02767a  4394b44c-b51d-5b96-b62b-db7f49007b43   \n",
       "4  7e1673f8-5758-4963-8804-d5e39f06205b  eebe9198-74c8-5996-bf88-c56d5fd6832f   \n",
       "\n",
       "                ethnicity  gender   race         submitter_id_demo  \\\n",
       "0  not hispanic or latino  female  white  TCGA-A2-A0YL_demographic   \n",
       "1  not hispanic or latino  female  white  TCGA-D8-A1XR_demographic   \n",
       "2  not hispanic or latino  female  white  TCGA-BH-A1ET_demographic   \n",
       "3  not hispanic or latino  female  white  TCGA-E2-A1LE_demographic   \n",
       "4  not hispanic or latino  female  asian  TCGA-C8-A12V_demographic   \n",
       "\n",
       "              updated_datetime_demo  year_of_birth  year_of_death  \\\n",
       "0  2016-09-02T19:00:43.039678-05:00         1962.0            NaN   \n",
       "1  2016-09-02T19:07:03.114741-05:00         1954.0            NaN   \n",
       "2  2016-09-02T18:57:52.265307-05:00         1944.0         2005.0   \n",
       "3  2016-09-02T19:09:17.163541-05:00         1937.0         2010.0   \n",
       "4  2016-09-02T19:02:30.147198-05:00         1955.0            NaN   \n",
       "\n",
       "   age_at_diagnosis                ...                 \\\n",
       "0           17702.0                ...                  \n",
       "1           20488.0                ...                  \n",
       "2           20425.0                ...                  \n",
       "3           26239.0                ...                  \n",
       "4           20346.0                ...                  \n",
       "\n",
       "  site_of_resection_or_biopsy       submitter_id_diag  \\\n",
       "0                       c50.9  TCGA-A2-A0YL_diagnosis   \n",
       "1                       c50.9  TCGA-D8-A1XR_diagnosis   \n",
       "2                       c50.9  TCGA-BH-A1ET_diagnosis   \n",
       "3                       c50.9  TCGA-E2-A1LE_diagnosis   \n",
       "4                       c50.9  TCGA-C8-A12V_diagnosis   \n",
       "\n",
       "   tissue_or_organ_of_origin   tumor_grade tumor_stage  \\\n",
       "0                      c50.9  not reported  stage iiia   \n",
       "1                      c50.9  not reported   stage iib   \n",
       "2                      c50.9  not reported     stage i   \n",
       "3                      c50.9  not reported  stage iiic   \n",
       "4                      c50.9  not reported   stage iia   \n",
       "\n",
       "              updated_datetime_diag vital_status  \\\n",
       "0  2016-09-02T19:00:43.039678-05:00        alive   \n",
       "1  2016-09-02T19:07:03.114741-05:00        alive   \n",
       "2  2016-09-02T18:57:52.265307-05:00         dead   \n",
       "3  2016-09-02T19:09:17.163541-05:00         dead   \n",
       "4  2016-09-02T19:02:30.147198-05:00        alive   \n",
       "\n",
       "                            exposure_id           submitter_id  \\\n",
       "0  23b6a296-e149-5801-aa20-e144ef683abe  TCGA-A2-A0YL_exposure   \n",
       "1  6dfa11dc-4266-57c9-b746-02a6c7bb9d23  TCGA-D8-A1XR_exposure   \n",
       "2  e72e6b02-58be-539a-af7f-f2714f14899a  TCGA-BH-A1ET_exposure   \n",
       "3  4c456a08-ffda-507f-a3e3-d273e5c49864  TCGA-E2-A1LE_exposure   \n",
       "4  dd5fbf53-64a4-5bc0-8ca7-ae05abbcbbac  TCGA-C8-A12V_exposure   \n",
       "\n",
       "                   updated_datetime  \n",
       "0  2016-09-02T19:00:43.039678-05:00  \n",
       "1  2016-09-02T19:07:03.114741-05:00  \n",
       "2  2016-09-02T18:57:52.265307-05:00  \n",
       "3  2016-09-02T19:09:17.163541-05:00  \n",
       "4  2016-09-02T19:02:30.147198-05:00  \n",
       "\n",
       "[5 rows x 30 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_clin.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Distances matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Creating DataFrame from the distance-correlation matrix data\n",
    "df_distances = pd.read_csv(DIST_PATH,sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA.3C.AAAU.0.Dis</th>\n",
       "      <th>TCGA.3C.AALI.0.Dis</th>\n",
       "      <th>TCGA.3C.AALJ.0.Dis</th>\n",
       "      <th>TCGA.3C.AALK.0.Dis</th>\n",
       "      <th>TCGA.4H.AAAK.0.Dis</th>\n",
       "      <th>TCGA.5L.AAT0.0.Dis</th>\n",
       "      <th>TCGA.5L.AAT1.0.Dis</th>\n",
       "      <th>TCGA.5T.A9QA.0.Dis</th>\n",
       "      <th>TCGA.A1.A0SB.0.Dis</th>\n",
       "      <th>TCGA.A1.A0SD.0.Dis</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA.UL.AAZ6.0.Dis</th>\n",
       "      <th>TCGA.UU.A93S.0.Dis</th>\n",
       "      <th>TCGA.V7.A7HQ.0.Dis</th>\n",
       "      <th>TCGA.W8.A86G.0.Dis</th>\n",
       "      <th>TCGA.WT.AB41.0.Dis</th>\n",
       "      <th>TCGA.WT.AB44.0.Dis</th>\n",
       "      <th>TCGA.XX.A899.0.Dis</th>\n",
       "      <th>TCGA.XX.A89A.0.Dis</th>\n",
       "      <th>TCGA.Z7.A8R5.0.Dis</th>\n",
       "      <th>TCGA.Z7.A8R6.0.Dis</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AAAU.0.Dis</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.120880</td>\n",
       "      <td>0.105801</td>\n",
       "      <td>0.109540</td>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.116020</td>\n",
       "      <td>0.138326</td>\n",
       "      <td>0.139012</td>\n",
       "      <td>0.164870</td>\n",
       "      <td>0.109789</td>\n",
       "      <td>...</td>\n",
       "      <td>0.130849</td>\n",
       "      <td>0.177643</td>\n",
       "      <td>0.202318</td>\n",
       "      <td>0.118387</td>\n",
       "      <td>0.204810</td>\n",
       "      <td>0.172970</td>\n",
       "      <td>0.131472</td>\n",
       "      <td>0.147361</td>\n",
       "      <td>0.160945</td>\n",
       "      <td>0.102312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AALI.0.Dis</th>\n",
       "      <td>0.120880</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.105240</td>\n",
       "      <td>0.102997</td>\n",
       "      <td>0.095894</td>\n",
       "      <td>0.142688</td>\n",
       "      <td>0.183251</td>\n",
       "      <td>0.114400</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117577</td>\n",
       "      <td>0.143124</td>\n",
       "      <td>0.172721</td>\n",
       "      <td>0.125615</td>\n",
       "      <td>0.158141</td>\n",
       "      <td>0.134712</td>\n",
       "      <td>0.116518</td>\n",
       "      <td>0.119447</td>\n",
       "      <td>0.129105</td>\n",
       "      <td>0.101502</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AALJ.0.Dis</th>\n",
       "      <td>0.105801</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.072653</td>\n",
       "      <td>0.091719</td>\n",
       "      <td>0.082123</td>\n",
       "      <td>0.096455</td>\n",
       "      <td>0.141006</td>\n",
       "      <td>0.181756</td>\n",
       "      <td>0.102686</td>\n",
       "      <td>...</td>\n",
       "      <td>0.108979</td>\n",
       "      <td>0.147735</td>\n",
       "      <td>0.151162</td>\n",
       "      <td>0.113216</td>\n",
       "      <td>0.138139</td>\n",
       "      <td>0.120631</td>\n",
       "      <td>0.100941</td>\n",
       "      <td>0.103558</td>\n",
       "      <td>0.112343</td>\n",
       "      <td>0.083557</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.3C.AALK.0.Dis</th>\n",
       "      <td>0.109540</td>\n",
       "      <td>0.086734</td>\n",
       "      <td>0.072653</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.040750</td>\n",
       "      <td>0.040376</td>\n",
       "      <td>0.062932</td>\n",
       "      <td>0.144121</td>\n",
       "      <td>0.129292</td>\n",
       "      <td>0.053960</td>\n",
       "      <td>...</td>\n",
       "      <td>0.117827</td>\n",
       "      <td>0.158452</td>\n",
       "      <td>0.146738</td>\n",
       "      <td>0.062870</td>\n",
       "      <td>0.165742</td>\n",
       "      <td>0.102312</td>\n",
       "      <td>0.060814</td>\n",
       "      <td>0.073400</td>\n",
       "      <td>0.080130</td>\n",
       "      <td>0.070721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TCGA.4H.AAAK.0.Dis</th>\n",
       "      <td>0.118014</td>\n",
       "      <td>0.105240</td>\n",
       "      <td>0.091719</td>\n",
       "      <td>0.040750</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.037074</td>\n",
       "      <td>0.065612</td>\n",
       "      <td>0.152471</td>\n",
       "      <td>0.113901</td>\n",
       "      <td>0.054209</td>\n",
       "      <td>...</td>\n",
       "      <td>0.143560</td>\n",
       "      <td>0.180884</td>\n",
       "      <td>0.172347</td>\n",
       "      <td>0.065300</td>\n",
       "      <td>0.187800</td>\n",
       "      <td>0.112343</td>\n",
       "      <td>0.059381</td>\n",
       "      <td>0.069163</td>\n",
       "      <td>0.097015</td>\n",
       "      <td>0.078198</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 1093 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    TCGA.3C.AAAU.0.Dis  TCGA.3C.AALI.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.000000            0.120880   \n",
       "TCGA.3C.AALI.0.Dis            0.120880            0.000000   \n",
       "TCGA.3C.AALJ.0.Dis            0.105801            0.086734   \n",
       "TCGA.3C.AALK.0.Dis            0.109540            0.086734   \n",
       "TCGA.4H.AAAK.0.Dis            0.118014            0.105240   \n",
       "\n",
       "                    TCGA.3C.AALJ.0.Dis  TCGA.3C.AALK.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.105801            0.109540   \n",
       "TCGA.3C.AALI.0.Dis            0.086734            0.086734   \n",
       "TCGA.3C.AALJ.0.Dis            0.000000            0.072653   \n",
       "TCGA.3C.AALK.0.Dis            0.072653            0.000000   \n",
       "TCGA.4H.AAAK.0.Dis            0.091719            0.040750   \n",
       "\n",
       "                    TCGA.4H.AAAK.0.Dis  TCGA.5L.AAT0.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.118014            0.116020   \n",
       "TCGA.3C.AALI.0.Dis            0.105240            0.102997   \n",
       "TCGA.3C.AALJ.0.Dis            0.091719            0.082123   \n",
       "TCGA.3C.AALK.0.Dis            0.040750            0.040376   \n",
       "TCGA.4H.AAAK.0.Dis            0.000000            0.037074   \n",
       "\n",
       "                    TCGA.5L.AAT1.0.Dis  TCGA.5T.A9QA.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.138326            0.139012   \n",
       "TCGA.3C.AALI.0.Dis            0.095894            0.142688   \n",
       "TCGA.3C.AALJ.0.Dis            0.096455            0.141006   \n",
       "TCGA.3C.AALK.0.Dis            0.062932            0.144121   \n",
       "TCGA.4H.AAAK.0.Dis            0.065612            0.152471   \n",
       "\n",
       "                    TCGA.A1.A0SB.0.Dis  TCGA.A1.A0SD.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.164870            0.109789   \n",
       "TCGA.3C.AALI.0.Dis            0.183251            0.114400   \n",
       "TCGA.3C.AALJ.0.Dis            0.181756            0.102686   \n",
       "TCGA.3C.AALK.0.Dis            0.129292            0.053960   \n",
       "TCGA.4H.AAAK.0.Dis            0.113901            0.054209   \n",
       "\n",
       "                           ...          TCGA.UL.AAZ6.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis         ...                    0.130849   \n",
       "TCGA.3C.AALI.0.Dis         ...                    0.117577   \n",
       "TCGA.3C.AALJ.0.Dis         ...                    0.108979   \n",
       "TCGA.3C.AALK.0.Dis         ...                    0.117827   \n",
       "TCGA.4H.AAAK.0.Dis         ...                    0.143560   \n",
       "\n",
       "                    TCGA.UU.A93S.0.Dis  TCGA.V7.A7HQ.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.177643            0.202318   \n",
       "TCGA.3C.AALI.0.Dis            0.143124            0.172721   \n",
       "TCGA.3C.AALJ.0.Dis            0.147735            0.151162   \n",
       "TCGA.3C.AALK.0.Dis            0.158452            0.146738   \n",
       "TCGA.4H.AAAK.0.Dis            0.180884            0.172347   \n",
       "\n",
       "                    TCGA.W8.A86G.0.Dis  TCGA.WT.AB41.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.118387            0.204810   \n",
       "TCGA.3C.AALI.0.Dis            0.125615            0.158141   \n",
       "TCGA.3C.AALJ.0.Dis            0.113216            0.138139   \n",
       "TCGA.3C.AALK.0.Dis            0.062870            0.165742   \n",
       "TCGA.4H.AAAK.0.Dis            0.065300            0.187800   \n",
       "\n",
       "                    TCGA.WT.AB44.0.Dis  TCGA.XX.A899.0.Dis  \\\n",
       "TCGA.3C.AAAU.0.Dis            0.172970            0.131472   \n",
       "TCGA.3C.AALI.0.Dis            0.134712            0.116518   \n",
       "TCGA.3C.AALJ.0.Dis            0.120631            0.100941   \n",
       "TCGA.3C.AALK.0.Dis            0.102312            0.060814   \n",
       "TCGA.4H.AAAK.0.Dis            0.112343            0.059381   \n",
       "\n",
       "                    TCGA.XX.A89A.0.Dis  TCGA.Z7.A8R5.0.Dis  TCGA.Z7.A8R6.0.Dis  \n",
       "TCGA.3C.AAAU.0.Dis            0.147361            0.160945            0.102312  \n",
       "TCGA.3C.AALI.0.Dis            0.119447            0.129105            0.101502  \n",
       "TCGA.3C.AALJ.0.Dis            0.103558            0.112343            0.083557  \n",
       "TCGA.3C.AALK.0.Dis            0.073400            0.080130            0.070721  \n",
       "TCGA.4H.AAAK.0.Dis            0.069163            0.097015            0.078198  \n",
       "\n",
       "[5 rows x 1093 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_distances.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Genome data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Creating DataFrame from the Genome data\n",
    "#Note please change the first two words of the text file: \"Hybridization REF\" to \"Hybridization_REF\"\n",
    "with open(GENOME_PATH) as f:\n",
    "    row = [row.split()[0:] for row in f]\n",
    "df_genome = pd.DataFrame(row[1:],columns=row[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Hybridization_REF</th>\n",
       "      <th>TCGA-3C-AAAU-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALI-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALJ-01A-31R-A41B-07</th>\n",
       "      <th>TCGA-3C-AALK-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-4H-AAAK-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5L-AAT0-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5L-AAT1-01A-12R-A41B-07</th>\n",
       "      <th>TCGA-5T-A9QA-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-A1-A0SB-01A-11R-A144-07</th>\n",
       "      <th>...</th>\n",
       "      <th>TCGA-UL-AAZ6-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-UU-A93S-01A-21R-A41B-07</th>\n",
       "      <th>TCGA-V7-A7HQ-01A-11R-A33J-07</th>\n",
       "      <th>TCGA-W8-A86G-01A-21R-A36F-07</th>\n",
       "      <th>TCGA-WT-AB41-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-WT-AB44-01A-11R-A41B-07</th>\n",
       "      <th>TCGA-XX-A899-01A-11R-A36F-07</th>\n",
       "      <th>TCGA-XX-A89A-01A-11R-A36F-07</th>\n",
       "      <th>TCGA-Z7-A8R5-01A-42R-A41B-07</th>\n",
       "      <th>TCGA-Z7-A8R6-01A-11R-A41B-07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>gene_id</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>...</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>?|100130426</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.9066</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>?|100133144</td>\n",
       "      <td>16.3644</td>\n",
       "      <td>9.2659</td>\n",
       "      <td>11.6228</td>\n",
       "      <td>12.0894</td>\n",
       "      <td>6.8468</td>\n",
       "      <td>3.9889</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>1.4644</td>\n",
       "      <td>15.3396</td>\n",
       "      <td>...</td>\n",
       "      <td>0.3992</td>\n",
       "      <td>4.3126</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>5.5624</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>14.3858</td>\n",
       "      <td>22.3240</td>\n",
       "      <td>2.2638</td>\n",
       "      <td>6.8865</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 1213 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Hybridization_REF TCGA-3C-AAAU-01A-11R-A41B-07 TCGA-3C-AALI-01A-11R-A41B-07  \\\n",
       "0           gene_id             normalized_count             normalized_count   \n",
       "1       ?|100130426                       0.0000                       0.0000   \n",
       "2       ?|100133144                      16.3644                       9.2659   \n",
       "\n",
       "  TCGA-3C-AALJ-01A-31R-A41B-07 TCGA-3C-AALK-01A-11R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.9066                       0.0000   \n",
       "2                      11.6228                      12.0894   \n",
       "\n",
       "  TCGA-4H-AAAK-01A-12R-A41B-07 TCGA-5L-AAT0-01A-12R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       6.8468                       3.9889   \n",
       "\n",
       "  TCGA-5L-AAT1-01A-12R-A41B-07 TCGA-5T-A9QA-01A-11R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.0000                       1.4644   \n",
       "\n",
       "  TCGA-A1-A0SB-01A-11R-A144-07             ...               \\\n",
       "0             normalized_count             ...                \n",
       "1                       0.0000             ...                \n",
       "2                      15.3396             ...                \n",
       "\n",
       "  TCGA-UL-AAZ6-01A-11R-A41B-07 TCGA-UU-A93S-01A-21R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.3992                       4.3126   \n",
       "\n",
       "  TCGA-V7-A7HQ-01A-11R-A33J-07 TCGA-W8-A86G-01A-21R-A36F-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.0000                       5.5624   \n",
       "\n",
       "  TCGA-WT-AB41-01A-11R-A41B-07 TCGA-WT-AB44-01A-11R-A41B-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                       0.0000                       0.0000   \n",
       "\n",
       "  TCGA-XX-A899-01A-11R-A36F-07 TCGA-XX-A89A-01A-11R-A36F-07  \\\n",
       "0             normalized_count             normalized_count   \n",
       "1                       0.0000                       0.0000   \n",
       "2                      14.3858                      22.3240   \n",
       "\n",
       "  TCGA-Z7-A8R5-01A-42R-A41B-07 TCGA-Z7-A8R6-01A-11R-A41B-07  \n",
       "0             normalized_count             normalized_count  \n",
       "1                       0.0000                       0.0000  \n",
       "2                       2.2638                       6.8865  \n",
       "\n",
       "[3 rows x 1213 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genome.head(n=3)\n",
    "#Row 0 as it has no information."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Removing data of the patients that are not in the three datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of references: 1093\n"
     ]
    }
   ],
   "source": [
    "#We know that the label \"TCGA-XX-XXXX\" is a reference for each patient\n",
    "#We take as a reference the file with less patients, that is to say: distances matrix data\n",
    "\n",
    "#Taking df_distances as reference to match the other datasets \n",
    "ref_distance = [val.replace(\".\",\"-\")[0:12] for val in df_distances.columns.tolist()]\n",
    "print(\"Number of references: \" + str(len(ref_distance)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of references: 1097\n",
      "set(['TCGA-AR-A0U1', 'TCGA-C8-A9FZ', 'TCGA-AC-A5EI', 'TCGA-A7-A0DC'])\n"
     ]
    }
   ],
   "source": [
    "#Getting the references for clinical data\n",
    "ref_clinical = [val[0:12] for val in df_clin['submitter_id_demo'].astype(str).tolist()]\n",
    "print(\"Number of references: \" + str(len(ref_clinical)))\n",
    "\n",
    "#Getting the references that are not in the matrix data (References that need to be removed)\n",
    "print(set(ref_clinical)-set(ref_distance))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of references: 1212\n",
      "set([])\n"
     ]
    }
   ],
   "source": [
    "#Getting the references for genome data\n",
    "ref_genome = [val[0:12]for val in df_genome.columns.tolist()[1:]]\n",
    "print(\"Number of references: \" + str(len(ref_genome)))\n",
    "\n",
    "#Getting the references that are not in the matrix data (References that need to be removed)\n",
    "print(set(ref_genome)-set(ref_distance))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 119 duplicates of references in genome data\n"
     ]
    }
   ],
   "source": [
    "#Since we saw that there are no extra references, we know then that there are duplicates.\n",
    "#Getting the duplicated data\n",
    "duplicates = [x for n, x in enumerate(ref_genome) if x in ref_genome[:n]]\n",
    "print(\"There are \" + str(len(duplicates)) + \" duplicated references in genome data\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['TCGA-A7-A0CE-11A-21R-A089-07',\n",
       " 'TCGA-A7-A0CE-01A-11R-A00Z-07',\n",
       " 'TCGA-A7-A0CH-11A-32R-A089-07',\n",
       " 'TCGA-A7-A0CH-01A-21R-A00Z-07']"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#These are the full name of the duplicates\n",
    "duplicates_full = [i for e in duplicates for i in df_genome.columns.tolist() if e in i]\n",
    "\n",
    "#Displaying just 2 duplicated references\n",
    "duplicates_full[0:4]"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "We see that both have the same reference that we were taking into acount, however, the rest of the characters are not the same and neither the values for the genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TCGA-A7-A0CE-11A-21R-A089-07</th>\n",
       "      <th>TCGA-A7-A0CE-01A-11R-A00Z-07</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>normalized_count</td>\n",
       "      <td>normalized_count</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>4.3333</td>\n",
       "      <td>11.0789</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3.9206</td>\n",
       "      <td>3.4986</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>78.9238</td>\n",
       "      <td>130.2890</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  TCGA-A7-A0CE-11A-21R-A089-07 TCGA-A7-A0CE-01A-11R-A00Z-07\n",
       "0             normalized_count             normalized_count\n",
       "1                       0.0000                       0.0000\n",
       "2                       4.3333                      11.0789\n",
       "3                       3.9206                       3.4986\n",
       "4                      78.9238                     130.2890"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_genome[[duplicates_full[0],duplicates_full[1]]].head()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
